File: dailyQuestion/2020/2020-09/9-20/golang/solution.go
package main

import (
	"fmt"
)

func main() {
	result := subsets([]int{2, 1, 3})
	fmt.Println(result)
}

func subsets(nums []int) (result [][]int) {
	temp := []int{}
	var dfs func(idx int)
	n := len(nums)
	dfs = func(idx int) {
		if idx == n {
			result = append(result, append([]int(nil), temp...))
			return
		}
		temp = append(temp, nums[idx])
		dfs(idx + 1)
		temp = temp[:len(temp)-1]
		dfs(idx + 1)
	}
	dfs(0)
	return
}

File: dailyQuestion/2020/2020-09/9-20/python/solution.py
def subsets(nums: [int]) -> [[int]]:
    result = []
    temp = []
    n = len(nums)
    def dfs(idx) :
        if idx == n :
            result.append(temp[:])
            return
        temp.append(nums[idx])
        dfs(idx +1)
        temp.pop()
        dfs(idx+1)
    dfs(0)
    return result


if __name__ == "__main__" :
    result = subsets([1,2,3])
    print(result)